src/backend/distributed/executor/executor.c

```diff
@@ -994,6 +1005,149 @@ FindOrCreateWorkerSession(WorkerPool *workerPool, MultiConnection *connection)
+
+		/*
+		 * Use some fields as the same which are safe to and required to
+		 * share across single task executions.
+		 */
+		singleTaskExecution->raiseInterrupts = execution->raiseInterrupts;
+		singleTaskExecution->isTransaction = execution->isTransaction;
+		singleTaskExecution->executionStats = execution->executionStats;
+		singleTaskExecution->scanState = execution->scanState;
+
+		/* we only expect to use a single connection for a single task */
+		singleTaskExecution->targetPoolSize = 1;
+		singleTaskExecution->unfinishedTaskCount = 1;
+		singleTaskExecution->totalTaskCount = 1;
+
+		/* RunDistributedExecution() will set the sessions, be explict here */
+		singleTaskExecution->sessionList = NIL;
+		singleTaskExecution->waitEventSet = NULL;
+
+		/* we're going to execute a single task only */
+		singleTaskExecution->plan = singleTaskDistributedPlan;
+
+		/*
+		 * StartDistributedExecution is common for all the executions (including
+		 * the parent execution) and so we don't need to re-call it for each
+		 * individual singleTaskExecution execution here.
+		 */
+
+		/* simply call the regular execution function */
+		RunDistributedExecution(singleTaskExecution);
+
+		/* remember the total number of rows processed */
+		execution->rowsProcessed += singleTaskExecution->rowsProcessed;
+
+		FinishDistributedExecution(singleTaskExecution);
+	}
+}
+
+
+/*
+ * CreateSingleTaskDistributedPlan gets a task and a distributed plan. The
+ * function returns a newly allocated distributed plan where the new plan
+ * contains only the task that is provided to the function along with
+ * some of the necessary fields set.
+ */
+static DistributedPlan *
+CreateSingleTaskDistributedPlan(Task *task, DistributedPlan *originalPlan)
+{
+	DistributedPlan *singleTaskPlan = palloc0(sizeof(DistributedPlan));
+
+	/* only copy the necessary parts for a single task execution */
+	singleTaskPlan->workerJob = copyObject(originalPlan->workerJob);
+	singleTaskPlan->workerJob->taskList = list_make1(task);
+
+	singleTaskPlan->operation = originalPlan->operation;
+	singleTaskPlan->hasReturning = originalPlan->hasReturning;
+	singleTaskPlan->targetRelationId = originalPlan->targetRelationId;
+
+	/* shallow copy of relationIdList is good enough */
+	singleTaskPlan->relationIdList = list_copy(originalPlan->relationIdList);
+
+	/*
+	 * We explicitly assert the fields to be NULL because we implemented the single task
+	 * plans for a restricted set of operations (see ShouldRunTasksSequentially()).
+	 * So, we don't expect other plans go through this code, unless implemented properly.
+	 */
+	Assert (singleTaskPlan->operation == CMD_INSERT);
+	Assert (singleTaskPlan->subPlanList == NIL);
+	Assert (singleTaskPlan->insertSelectSubquery == NULL);
+	Assert (singleTaskPlan->insertTargetList == NIL);
+	Assert (singleTaskPlan->intermediateResultIdPrefix == NULL);
+	Assert (singleTaskPlan->masterQuery == NULL);
+	Assert (list_length(originalPlan->workerJob->taskList) > 1);
```

Comment:
It is not really clear from the name of the function that these `Assert()`s should be done here. Also this creates a tight coupling between `ShouldRunTasksSequentially()` and this function (i.e. whenever we modify one, we should also modify the other), while it doesn't seem very necessary.

What do you think about moving these `Assert()`s to the `if (initialTask->rowValuesLists != NIL)` block in `ShouldRunTasksSequentially()`?